autopilot/hostscore.go

```diff
@@ -0,0 +1,241 @@
+package autopilot
+
+import (
+	"math"
+	"math/big"
+	"sort"
+	"time"
+
+	"go.sia.tech/renterd/hostdb"
+	"go.sia.tech/siad/build"
+	"go.sia.tech/siad/types"
+	"lukechampine.com/frand"
+)
+
+const (
+	// maxBaseRPCPriceVsBandwidth is the max ratio for sane pricing between the
+	// MinBaseRPCPrice and the MinDownloadBandwidthPrice. This ensures that 1
+	// million base RPC charges are at most 1% of the cost to download 4TB. This
+	// ratio should be used by checking that the MinBaseRPCPrice is less than or
+	// equal to the MinDownloadBandwidthPrice multiplied by this constant
+	maxBaseRPCPriceVsBandwidth = uint64(40e3)
+
+	// maxSectorAccessPriceVsBandwidth is the max ratio for sane pricing between
+	// the MinSectorAccessPrice and the MinDownloadBandwidthPrice. This ensures
+	// that 1 million base accesses are at most 10% of the cost to download 4TB.
+	// This ratio should be used by checking that the MinSectorAccessPrice is
+	// less than or equal to the MinDownloadBandwidthPrice multiplied by this
+	// constant
+	maxSectorAccessPriceVsBandwidth = uint64(400e3)
+)
+
+type scoreFn func(h host) float64
+
+func HostScore(scoreFns ...scoreFn) scoreFn {
```

Comment:
I don't think this structure makes as much sense for scoring. The hosts score always consists of all the individual parts so we should never pass in fewer than all of them here.

Reply:
I decided to mirror host filtering for two reasons: because it's nice having things behave in the same way but also because I figured we _would_ run into scenarios where you don't necessarily want to include all scoring functions. 

Like why would `uptime`, `interactions`, `age` and `storageRemaining` be taken up when scoring new hosts.
I can see why including them probably doesn't hurt but isn't plainly excluding them way better? E.g. if we had a contract with a host once for a limited amount of time and then dropped it, (so probably not a great host) it would have an edge over other hosts because it would have an `age` advantage.

I'm looking for reasons to keep it the way it is though. I'm fine with getting rid of it I don't feel that strongly.
Don't really see the point though if I refactor it to be an accumulator. We can easily add an extra method that accumulates all scoring functions and use that, leaving the flexibility should we ever need it.